src/common/com/intellij/plugins/haxe/ide/folding/HaxeFoldingBuilder.java

```diff
@@ -22,37 +23,138 @@
+import static com.intellij.plugins.haxe.lang.lexer.HaxeTokenTypes.*;
+import static com.intellij.psi.TokenType.WHITE_SPACE;
+
 public class HaxeFoldingBuilder implements FoldingBuilder {
+
+  public static final String PLACEHOLDER_TEXT = "...";
+
   @NotNull
   @Override
   public FoldingDescriptor[] buildFoldRegions(@NotNull ASTNode node, @NotNull Document document) {
-    List<FoldingDescriptor> list = new ArrayList<FoldingDescriptor>();
+    List<FoldingDescriptor> list = new ArrayList<>();
     buildFolding(node, list);
     FoldingDescriptor[] descriptors = new FoldingDescriptor[list.size()];
     return list.toArray(descriptors);
   }
 
-  private static void buildFolding(ASTNode node, List<FoldingDescriptor> list) {
-    if (node.getElementType() == HaxeTokenTypes.BLOCK_STATEMENT) {
-      final TextRange range = node.getTextRange();
-      list.add(new FoldingDescriptor(node, range));
+  private static void buildFolding(@NotNull ASTNode node, List<FoldingDescriptor> list) {
+    final IElementType elementType = node.getElementType();
+
+    FoldingDescriptor descriptor = null;
+    if (isImportOrUsingStatement(elementType) && isFirstImportStatement(node)) {
+      descriptor = buildImportsFolding(node);
+    } else if (isCodeBlock(elementType)) {
+      descriptor = buildCodeBlockFolding(node);
+    } else if (isBodyBlock(elementType)) {
+      descriptor = buildBodyBlockFolding(node);
+    }
+
+    if (descriptor != null) {
+      list.add(descriptor);
     }
+
     for (ASTNode child : node.getChildren(null)) {
       buildFolding(child, list);
     }
   }
 
+  private static boolean isCodeBlock(IElementType elementType) {
+    return elementType == BLOCK_STATEMENT;
+  }
+
+  private static boolean isBodyBlock(IElementType elementType) {
+    return BODY_TYPES.contains(elementType);
+  }
+
+  private static FoldingDescriptor buildCodeBlockFolding(@NotNull ASTNode node) {
+    final ASTNode openBrace = node.getFirstChildNode();
+    final ASTNode closeBrace = node.getLastChildNode();
+
+    return buildBlockFolding(node, openBrace, closeBrace);
+  }
+
+  private static FoldingDescriptor buildBodyBlockFolding(@NotNull ASTNode node) {
+    final ASTNode openBrace = UsefulPsiTreeUtil.getPrevSiblingSkipWhiteSpacesAndComments(node);
+    final ASTNode closeBrace = UsefulPsiTreeUtil.getNextSiblingSkipWhiteSpacesAndComments(node);
+
+    return buildBlockFolding(node, openBrace, closeBrace);
+  }
+
+  private static FoldingDescriptor buildBlockFolding(@NotNull ASTNode node, ASTNode openBrace, ASTNode closeBrace) {
+    TextRange textRange;
+    if (openBrace != null && closeBrace != null && openBrace.getElementType() == PLCURLY && closeBrace.getElementType() == PRCURLY) {
+      textRange = new TextRange(openBrace.getTextRange().getEndOffset(), closeBrace.getStartOffset());
+    } else {
+      textRange = node.getTextRange();
+    }
+
+    if (textRange.getLength() > PLACEHOLDER_TEXT.length()) {
```

Comment:
Jut a thought for a refinement: We might want to check if we can make a placeholder that's a one-liner (<screenWidth chars) that gives full context info on screen.  Like this:
```haxe
  public static function doIt() {
    trace(1);
  }
```
becomes
```haxe
  public static function doIt() { trace(1); }
```
rather than
```haxe
  public static function doIt() { ... }
```

Reply:
https://github.com/HaxeFoundation/intellij-haxe/issues/737